### DISCUSSION TITLE
DISTINCT ON with ORDERBY - inconsistent behaviour from documentation

### DISCUSSION THREAD

**Discussion by ninist** [ClickHouse/ClickHouse#62440] [Q&A]
https://clickhouse.com/docs/en/sql-reference/statements/select/distinct
from issue #25404 and pull request #25589

The documentation is fairly brief on `SELECT DISTINCT ON (col1,col2)`, in particular it does not say whether order matters when the distinct-relation is applied.

The section in the documentation on `DISTINCT and ORDER BY`, however, mentions that 
>    The `DISTINCT` clause is executed before the `ORDER BY` clause.

This does not appear to be the case for `DISTINCT ON`:

```
create table tmp_distinct_table (
  ts UInt32,
  ver UInt32,
  value UInt32
) engine = MergeTree()
order by (ts, ver);

insert into tmp_distinct_table values
  (1, 1, 10),
  (1, 2, 20),
  (2, 1, 5),
  (3, 2, 30),
  (3, 1, 30),
  (4, 2, 40),
  (4, 1, 41),
  (4, 3, 42);
```
This table now contains
```
ts|ver|value|
--+---+-----+
 1|  1|   10|
 1|  2|   20|
 2|  1|    5|
 3|  1|   30|
 3|  2|   30|
 4|  1|   41|
 4|  2|   40|
 4|  3|   42|
```
Take the following query, aiming to select unique `ts` with the highest value for `ver`:
```
select distinct on (ts)
  ts,
  ver,
  value
from (
  select ts,ver,value
  from tmp_distinct_table
  order by ts,ver desc
)
order by ts;
```
The result is
```
ts|ver|value|
--+---+-----+
 1|  1|   10|
 2|  1|    5|
 3|  1|   30|
 4|  1|   41|
```
Which from `ver=1` may indicate that either
* It is arbitrary which rows are dropped to satisfy `DISTINCT ON(ts)`.
* The outer `ORDER BY`-clause was applied before applying `DISTINCT ON` (contrary to what the documentation claims for `DISTINCT`).

Take this second query (the only difference is appending `ver desc` at the very end in the last sort):
```
select distinct on (ts)
  ts,
  ver,
  value
from (
  select ts,ver,value
  from tmp_distinct_table
  order by ts,ver desc
)
order by ts,ver desc;
```

The result is
```
ts|ver|value|
--+---+-----+
 1|  2|   20|
 2|  1|    5|
 3|  2|   30|
 4|  3|   42|
```

This is the result I expected from my initial understanding of the documentation.

My questions are
* Does order matter for the `DISTINCT` and `DISTINCT ON` clauses?
* Should the documentation perhaps be updated to specify that `ORDER BY` is in fact applied before `DISTINCT ON`, even if this is not the case for `DISTINCT`?

```
select version();

version()|
---------+
23.2.4.12|
```

Very late edit: noticed a minor typo in one of the clarifying aside comments: `ver -> ts`, does not otherwise affect the issue/discussion.

**Comment by movy** [comment]
I can confirm this observation: sorting by a column NOT included in `DISTINCT ON` clause affects which rows will be included in 'distincted' output, examples are below, the only difference is sort direction for `note` column:

```sql
select
    distinct on (symbol) *
from
    backtests
order by
    symbol,
    note ASC
```

![image](https://github.com/user-attachments/assets/6e93bb14-e2d5-450c-929e-27423736b814)

```sql
select
    distinct on (symbol) *
from
    backtests
order by
    symbol,
    note DESC
```
![image](https://github.com/user-attachments/assets/62837199-94c9-4db0-809b-89c42813318c)

**Comment by pratimapatel2008** [comment]
@ninist Going by the documentation, it mentions about distinct clause where distinct by is executed first followed by order by, However in this case your outer query has DISTINCT ON (which is not same as distinct) clause which seems to be expected. The query can also be rewritten as below.

```
SELECT
    ts,
    ver,
    value
FROM
(
    SELECT
        ts,
        ver,
        value
    FROM tmp_distinct_table
    ORDER BY
        ts ASC,
        ver DESC
)
ORDER BY ts ASC
LIMIT 1 BY ts
```

I think for the results which you were expecting, the query can be written as
```
with a as (
  select ts,ver,value
  from tmp_distinct_table
  order by (ts,ver) DESC
),
B as (select distinct on (ts)
  ts,
  ver,
  value from a)
select * from B order by ts;
```

**Reply by ninist** [reply]
>However in this case your outer query has DISTINCT ON (which is not same as distinct)

Yes, that is what this discussion/issue is about - the documentation has a shared page for DISTINCT and DISTINCT ON, but the documentation is not explicit on the ordering for DISTINCT ON.

I believe my second query is equivalent to your first query (the parser converts it to that).

Is your second query the preferred way to write this?

In any case, I believe the documentation can be made more explicit here.

**Reply by czlee** [reply]
Do I understand right that `DISTINCT ON (ts)` is just an alias for `LIMIT 1 BY ts`? The discussion in #25404 seems to suggest this, but I don't think the documentation says this. Rather than going to the trouble of writing a new explanation in the DISTINCT page, would it be simpler just to say that `DISTINCT ON` is an alias for `LIMIT 1 BY`, and to link to the documentation on `LIMIT BY`? https://clickhouse.com/docs/en/sql-reference/statements/select/limit-by

**Reply by ninist** [reply]
That is my understanding too.

Anything that could make it more clear the section order by does not apply to `DISTINCT ON` would be good in my opinion, since the two distinct statements are currently somewhat intermingled in the `DISTINCT`-clause docs. 

As it currently is, I found it easy to make the (wrong) assumption that the section on `ORDER BY` applies to both, but that the examples and descriptions were not repeated for `DISTINCT ON` for the sake of verbosity.
